dot/parachain/backing/can_second.go

```diff
@@ -0,0 +1,141 @@
+	if !ok || rpState == nil {
+		// relay parent is unknown or relay parent state is nil.
+		msg.ResCh <- false
+		return
+	}
+
+	ppMode := rpState.prospectiveParachainsMode
+	if !ppMode.IsEnabled {
+		// async backing is disabled.
+		msg.ResCh <- false
+		return
+	}
+
+	hypotheticalCandidate := parachaintypes.HypotheticalCandidateIncomplete{
+		CandidateHash:      msg.CandidateHash,
+		CandidateParaID:    msg.CandidateParaID,
+		ParentHeadDataHash: msg.ParentHeadDataHash,
+		RelayParent:        msg.CandidateRelayParent,
+	}
+
+	isSecondingAllowed, membership := cb.secondingSanityCheck(hypotheticalCandidate, true)
+
+	if !isSecondingAllowed {
+		msg.ResCh <- false
+		return
+	}
+
+	for _, fragmentTree := range membership {
+		// candidate should be recognised by at least some fragment tree.
+		if len(fragmentTree) != 0 {
+			msg.ResCh <- true
+			return
+		}
+	}
+}
+
+// secondingSanityCheck checks whether a candidate can be seconded based on its
+// hypothetical frontiers in the fragment tree and what we've already seconded in
+// all active leaves.
+//
+// if the candidate can be seconded, returns true and a map of the heads of active leaves to the depths,
+// where the candidate is a member of the fragment tree.
+// Returns false if the candidate cannot be seconded.
+func (cb *CandidateBacking) secondingSanityCheck(
+	hypotheticalCandidate parachaintypes.HypotheticalCandidate,
+	backedInPathOnly bool, //nolint:unparam
+) (bool, map[common.Hash][]uint) {
+	type response struct {
+		depths          []uint
+		head            common.Hash
+		activeLeafState activeLeafState
+	}
+
+	var (
+		responses            []response
+		candidateParaID      parachaintypes.ParaID
+		candidateRelayParent common.Hash
+		membership           = make(map[common.Hash][]uint)
+	)
+
+	switch v := hypotheticalCandidate.(type) {
+	case parachaintypes.HypotheticalCandidateIncomplete:
+		candidateParaID = v.CandidateParaID
+		candidateRelayParent = v.RelayParent
+	case parachaintypes.HypotheticalCandidateComplete:
+		candidateParaID = parachaintypes.ParaID(v.CommittedCandidateReceipt.Descriptor.ParaID)
+		candidateRelayParent = v.CommittedCandidateReceipt.Descriptor.RelayParent
+	}
+
+	for head, leafState := range cb.perLeaf {
+		if leafState.prospectiveParachainsMode.IsEnabled {
+			allowedParentsForPara := cb.implicitView.knownAllowedRelayParentsUnder(head, candidateParaID)
+
+			if !slices.Contains(allowedParentsForPara, candidateRelayParent) {
+				continue
+			}
+
+			responseCh := make(chan parachaintypes.HypotheticalFrontierResponse)
+			cb.SubSystemToOverseer <- parachaintypes.ProspectiveParachainsMessageGetHypotheticalFrontier{
+				HypotheticalFrontierRequest: parachaintypes.HypotheticalFrontierRequest{
+					Candidates:              []parachaintypes.HypotheticalCandidate{hypotheticalCandidate},
+					FragmentTreeRelayParent: &head,
+					BackedInPathOnly:        backedInPathOnly,
+				},
+				Ch: responseCh,
+			}
```

Comment:
could you create a constructor for this struct?
```suggestion
			cb.SubSystemToOverseer <-  parachaintypes.
			    NewProspectiveParachainsMessage(responseCh).
			    GetHypotheticalFrontier(hypotheticalCandidate, &head, backedInPathOnly) 
```

And then in the `parachaintypes` package you can create:

```go
NewProspectiveParachainsMessage(responseCh chan HypotheticalFrontierResponse) *ProspectiveParachainsMessageGetHypotheticalFrontier {
    return &ProspectiveParachainsMessageGetHypotheticalFrontier{Ch: responseCh}
}

func (p *) GetHypotheticalFrontier(args ...) *ProspectiveParachainsMessageGetHypotheticalFrontier {
    p.HypotheticalFrontierRequest = HypotheticalFrontierRequest{
        Candidates: []HypotheticalCandidate{hypotheticalCandidate},
            FragmentTreeRelayParent: &head,
	    BackedInPathOnly: backedInPathOnly,
    }

    return p
}
```